README.md
[![Published on webcomponents.org](https://img.shields.io/badge/webcomponents.org-published-blue.svg)](https://www.webcomponents.org/element/hotforfeature/delightful-icons)
[![Commitizen friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](http://commitizen.github.io/cz-cli/)

_[Demo and API docs](https://www.webcomponents.org/element/hotforfeature/delightful-icons)_

## &lt;delightful-icons&gt;

Material Design: [Creative Customization](https://material.io/guidelines/motion/creative-customization.html#creative-customization-icons)

`delightful-icons` is a collection of icon sets and a helper `<delightful-icon>` element to assist transitioning between them. Each icon set is a pair of icons that can transition from one to another with an animation. These animations are part of Material Design's delightful details.

### &lt;delightful-icon&gt;

`<delightful-icon>` is a helper element that wraps any element that uses an `<iron-icon>`. You can use it to define which icons to transition to.

```html
<!-- Import transitions individually -->
<link rel="import" href="./bower_components/delightful-icons/icons/menu-arrow-back.html">
<link rel="import" href="./bower_components/delightful-icons/icons/menu-arrow-forward.html">

<!-- Or all transitions -->
<link rel="import" href="./bower_components/delightful-icons/icons/all.html">

<!-- Import transitions before <delightful-icon> element -->
<link rel="import" href="./bower_components/delightful-icons/delightful-icon.html">
```

<!---
```
<custom-element-demo>
  <template>
    <script src="../webcomponentsjs/webcomponents-lite.js"></script>
    <link rel="import" href="../paper-icon-button/paper-icon-button.html">
    <link rel="import" href="icons/add-create.html">
    <link rel="import" href="icons/menu-arrow-back.html">
    <link rel="import" href="icons/play-arrow-pause.html">
    <link rel="import" href="delightful-icon.html">
    <next-code-block></next-code-block>
  </template>
</custom-element-demo>
```
-->
```html
<delightful-icon icon="menu" next="arrow-back" toggle-on="click">
  <paper-icon-button></paper-icon-button>
</delightful-icon>

<delightful-icon icon="play-arrow" next="pause" toggle-on="click">
  <paper-icon-button></paper-icon-button>
</delightful-icon>

<delightful-icon icon="add" next="create" toggle-on="click">
  <paper-icon-button></paper-icon-button>
</delightful-icon>
```

There are two ways to trigger a transition:

1. Change the `icon` property
2. Set the `next` property and invoke `toggle()`

`<delightful-icon>` can listen to a default event from its wrapped element to toggle on by setting the `toggle-on` attribute to the event name.

#### Wrapped element

By default, `<delightful-icon>` will use its first child as the wrapped element and set the `icon` property. This can be customized further with `query` and `property` attributes.

```html
<delightful-icon icon="menu" next="arrow-back" toggle-on="click"
    query=".use-this" property="displayIcon">
  <div>
    <!-- <delightful-icon> will set the displayIcon property of this element -->
    <my-custom-element class="use-this"></my-custom-element>
  </div>
</delightful-icon>
```

#### Multi-icon transitions

You can transition between several icons, provided each icon transitioned to has an imported animation for the icon it is transitioning from.

<!---
```
<custom-element-demo>
  <template>
    <script src="../webcomponentsjs/webcomponents-lite.js"></script>
    <link rel="import" href="../paper-icon-button/paper-icon-button.html">
    <link rel="import" href="icons/menu-arrow-back.html">
    <link rel="import" href="icons/menu-arrow-forward.html">
    <link rel="import" href="icons/arrow-back-arrow-forward.html">
    <link rel="import" href="delightful-icon.html">
    <next-code-block></next-code-block>
  </template>
</custom-element-demo>
```
-->
```html
<!-- This button will transition between menu, arrow-back, and arrow-forward -->
<delightful-icon id="menu-arrows" icon="menu">
  <paper-icon-button></paper-icon-button>
</delightful-icon>

<script>
window.addEventListener('WebComponentsReady', function() {
  var delight = document.querySelector('#menu-arrows');
  delight.iconElement.addEventListener('click', function() {
    if (delight.icon === 'menu') {
      delight.icon = 'arrow-back';
    } else if (delight.icon === 'arrow-back') {
      delight.icon = 'arrow-forward';
    } else {
      delight.icon = 'menu';
    }
  });
});
</script>
```

### Styling

The following custom properties are available for styling:

Custom Property               | Description         | Default
------------------------------|---------------------|----------
`--delightful-icons-duration` | Transition duration | `0.2s`
`--delightful-icons-easing`   | Transition easing   | `cubic-bezier(0.4, 0, 0.2, 1)`

### Compatibility

On browsers that do not support CSS transforms, the icon will immediately change from one to another without a transition.

- **IE 11** does not support CSS transforms on SVG elements.
- [**Microsoft Edge** is considering CSS transforms for SVG](https://developer.microsoft.com/en-us/microsoft-edge/platform/status/supportcsstransformsonsvg/).
- **Safari 9** supports CSS transforms, but [due to other maintenance complications](https://github.com/hotforfeature/delightful-icons/issues/3) I have chosen to disable animations for 9. Safari 10+ will show animations correctly.



CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.5.0"></a>
# [0.5.0](https://github.com/hotforfeature/delightful-icons/compare/v0.4.0...v0.5.0) (2018-08-14)


### Features

* support Polymer 3 ([c17f84b](https://github.com/hotforfeature/delightful-icons/commit/c17f84b))



<a name="0.4.0"></a>
# [0.4.0](https://github.com/hotforfeature/delightful-icons/compare/v0.3.0...v0.4.0) (2017-11-14)


### Bug Fixes

* SVG transform feature detection ([13e161d](https://github.com/hotforfeature/delightful-icons/commit/13e161d)), closes [#4](https://github.com/hotforfeature/delightful-icons/issues/4)
* **icon:** disableTransition set to false when initialized with true ([6f5e0d0](https://github.com/hotforfeature/delightful-icons/commit/6f5e0d0))
* **icon:** disabling transitions on Safari not working ([bedabf9](https://github.com/hotforfeature/delightful-icons/commit/bedabf9))
* **icon:** remove ES6 use of Array.prototype.find ([7b0744d](https://github.com/hotforfeature/delightful-icons/commit/7b0744d))


### Documentation

* update readme to use helper element and move all icons import to icons folder ([36d5d5e](https://github.com/hotforfeature/delightful-icons/commit/36d5d5e))


### Features

* **icon:** add element to automatically select and toggle between icons ([e6070c3](https://github.com/hotforfeature/delightful-icons/commit/e6070c3)), closes [#2](https://github.com/hotforfeature/delightful-icons/issues/2)
* **icon:** add toggle-on attribute ([008a908](https://github.com/hotforfeature/delightful-icons/commit/008a908))
* **icons:** add arrow-back/arrow-forward ([c240128](https://github.com/hotforfeature/delightful-icons/commit/c240128))


### BREAKING CHANGES

* to import all icons, import "icon/all.html" instead of "all.html"



<a name="0.3.0"></a>
# [0.3.0](https://github.com/hotforfeature/delightful-icons/compare/v0.2.0...v0.3.0) (2017-11-02)


### Features

* add add/create pair ([e1433c7](https://github.com/hotforfeature/delightful-icons/commit/e1433c7)), closes [#5](https://github.com/hotforfeature/delightful-icons/issues/5)



<a name="0.2.0"></a>
# [0.2.0](https://github.com/hotforfeature/delightful-icons/compare/v0.1.1...v0.2.0) (2017-11-02)


### Bug Fixes

* Safari 9 icons not displaying ([bb0e039](https://github.com/hotforfeature/delightful-icons/commit/bb0e039)), closes [#3](https://github.com/hotforfeature/delightful-icons/issues/3)


### Features

* add play-arrow/pause pair ([65549a7](https://github.com/hotforfeature/delightful-icons/commit/65549a7)), closes [#1](https://github.com/hotforfeature/delightful-icons/issues/1)



<a name="0.1.1"></a>
## [0.1.1](https://github.com/hotforfeature/delightful-icons/compare/v0.1.0...v0.1.1) (2017-11-01)



<a name="0.1.0"></a>
# [0.1.0](https://github.com/hotforfeature/delightful-icons/compare/b62fda8...v0.1.0) (2017-10-31)


### Features

* **icons:** add menu/arrow-back and arrow-forward pairs ([b62fda8](https://github.com/hotforfeature/delightful-icons/commit/b62fda8))


icons/all.js
import './add-create.js';
import './arrow-back-arrow-forward.js';
import './menu-arrow-back.js';
import './menu-arrow-forward.js';
import './play-arrow-pause.js';

/**
@license
MIT License

Copyright (c) 2017 Andrew Mitchell

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
/*
  FIXME(polymer-modulizer): the above comments were extracted
  from HTML and may be out of place here. Review them and
  then delete this comment!
*/
;


icons/add-create.js
import '@polymer/iron-iconset-svg/iron-iconset-svg.js';
import '../delightful-icons-style.js';
const $_documentContainer = document.createElement('template');

$_documentContainer.innerHTML = `<iron-iconset-svg name="delightful-add-create" size="24">
  <svg>
    <defs>
      <g id="add" class="aca">
        <style type="text/css">
          @keyframes aca-w {
            from {
              transform: rotate(45deg);
              visibility: visible;
            }
            to {
              transform: rotate(-90deg);
              visibility: hidden;
            }
          }

          @keyframes aca-a {
            from {
              transform: translate(-2.65px, -5.55px) scale(5.3, 18.3);
              -webkit-clip-path: polygon(0 0, 100% 0, 100% 85.5%, 50% 100%, 0 85.5%);
              clip-path: polygon(0 0, 100% 0, 100% 85.5%, 50% 100%, 0 85.5%);
            }
            to {
              transform: translate(-1px, -7px) scale(2, 14);
              -webkit-clip-path: polygon(0 0, 100% 0, 100% 100%, 50% 100%, 0 100%);
              clip-path: polygon(0 0, 100% 0, 100% 100%, 50% 100%, 0 100%);
            }
          }

          @keyframes aca-b {
            from {
              transform: translate(-2.65px, -10.65px) scale(5.3, 3.55);
              -webkit-clip-path: polygon(0 25%, 1% 20%, 2% 15%, 4% 10%, 8% 5%, 12% 2%, 16% 0,
                84% 0, 88% 2%, 92% 5%, 96% 10%, 98% 15%, 99% 20%, 100% 25%,
                100% 100%, 0 100%);
              clip-path: polygon(0 25%, 1% 20%, 2% 15%, 4% 10%, 8% 5%, 12% 2%, 16% 0,
                84% 0, 88% 2%, 92% 5%, 96% 10%, 98% 15%, 99% 20%, 100% 25%,
                100% 100%, 0 100%);
            }
            to {
              transform: translate(-7px, -1px) scale(14, 2);
              -webkit-clip-path: polygon(0 0, 0 0, 0 0, 0 0, 0 0, 0 0, 0 0,
                100% 0, 100% 0, 100% 0, 100% 0, 100% 0, 100% 0, 100% 0,
                100% 100%, 0 100%);
              clip-path: polygon(0 0, 0 0, 0 0, 0 0, 0 0, 0 0, 0 0,
                100% 0, 100% 0, 100% 0, 100% 0, 100% 0, 100% 0, 100% 0,
                100% 100%, 0 100%);
            }
          }

          @keyframes aca-v {
            from {
              visibility: hidden;
            }
            to {
              visibility: visible;
            }
          }

          .aca .w,
          .aca .a,
          .aca .b {
            animation-duration: var(--delightful-icons-duration);
            animation-fill-mode: forwards;
            animation-timing-function: var(--delightful-icons-easing);
          }

          .aca .w {
            animation-name: aca-w;
          }

          .aca .a {
            animation-name: aca-a;
          }

          .aca .b {
            animation-name: aca-b;
          }

          .aca path {
            animation-name: aca-v;
            animation-delay: var(--delightful-icons-duration);
            animation-fill-mode: forwards;
            visibility: hidden;
          }
        </style>
        <g transform="translate(12, 12)" stroke="currentcolor">
          <g class="w">
            <rect class="a" x="0" y="0" width="1" height="1"></rect>
            <rect class="b" x="0" y="0" width="1" height="1"></rect>
          </g>
        </g>
        <path d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"></path>
      </g>

      <g id="create" class="acc">
        <style type="text/css">
          @keyframes acc-w {
            from {
              transform: rotate(-90deg);
              visibility: visible;
            }
            to {
              transform: rotate(45deg);
              visibility: hidden;
            }
          }

          @keyframes acc-a {
            from {
              transform: translate(-1px, -7px) scale(2, 14);
              -webkit-clip-path: polygon(0 0, 100% 0, 100% 100%, 50% 100%, 0 100%);
              clip-path: polygon(0 0, 100% 0, 100% 100%, 50% 100%, 0 100%);
            }
            to {
              transform: translate(-2.65px, -5.55px) scale(5.3, 18.3);
              -webkit-clip-path: polygon(0 0, 100% 0, 100% 85.5%, 50% 100%, 0 85.5%);
              clip-path: polygon(0 0, 100% 0, 100% 85.5%, 50% 100%, 0 85.5%);
            }
          }

          @keyframes acc-b {
            from {
              transform: translate(-7px, -1px) scale(14, 2);
              -webkit-clip-path: polygon(0 0, 0 0, 0 0, 0 0, 0 0, 0 0, 0 0,
                100% 0, 100% 0, 100% 0, 100% 0, 100% 0, 100% 0, 100% 0,
                100% 100%, 0 100%);
              clip-path: polygon(0 0, 0 0, 0 0, 0 0, 0 0, 0 0, 0 0,
                100% 0, 100% 0, 100% 0, 100% 0, 100% 0, 100% 0, 100% 0,
                100% 100%, 0 100%);
            }
            to {
              transform: translate(-2.65px, -10.65px) scale(5.3, 3.55);
              -webkit-clip-path: polygon(0 25%, 1% 20%, 2% 15%, 4% 10%, 8% 5%, 12% 2%, 16% 0,
                84% 0, 88% 2%, 92% 5%, 96% 10%, 98% 15%, 99% 20%, 100% 25%,
                100% 100%, 0 100%);
              clip-path: polygon(0 25%, 1% 20%, 2% 15%, 4% 10%, 8% 5%, 12% 2%, 16% 0,
                84% 0, 88% 2%, 92% 5%, 96% 10%, 98% 15%, 99% 20%, 100% 25%,
                100% 100%, 0 100%);
            }
          }

          @keyframes acc-v {
            from {
              visibility: hidden;
            }
            to {
              visibility: visible;
            }
          }

          .acc .w,
          .acc .a,
          .acc .b {
            animation-duration: var(--delightful-icons-duration);
            animation-fill-mode: forwards;
            animation-timing-function: var(--delightful-icons-easing);
          }

          .acc .w {
            animation-name: acc-w;
          }

          .acc .a {
            animation-name: acc-a;
          }

          .acc .b {
            animation-name: acc-b;
          }

          .acc path {
            animation-name: acc-v;
            animation-delay: var(--delightful-icons-duration);
            animation-fill-mode: forwards;
            visibility: hidden;
          }
        </style>
        <g transform="translate(12, 12)" stroke="currentcolor">
          <g class="w">
            <rect class="a" x="0" y="0" width="1" height="1"></rect>
            <rect class="b" x="0" y="0" width="1" height="1"></rect>
          </g>
        </g>
        <path d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34c-.39-.39-1.02-.39-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"></path>
      </g>
    </defs>
  </svg>
</iron-iconset-svg>`;

document.head.appendChild($_documentContainer.content);

/**
@license
MIT License

Copyright (c) 2017 Andrew Mitchell

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
/*
  FIXME(polymer-modulizer): the above comments were extracted
  from HTML and may be out of place here. Review them and
  then delete this comment!
*/
;


icons/menu-arrow-back.js
import '@polymer/iron-iconset-svg/iron-iconset-svg.js';
import '../delightful-icons-style.js';
const $_documentContainer = document.createElement('template');

$_documentContainer.innerHTML = `<iron-iconset-svg name="delightful-menu-arrow-back" size="24">
  <svg>
    <defs>
      <g id="arrow-back" class="mabab">
        <style type="text/css">
          @keyframes mabab-w {
            from {
              transform: rotate(0deg);
              visibility: visible;
            }
            to {
              transform: rotate(180deg);
              visibility: hidden;
            }
          }

          @keyframes mabab-a {
            from {
              transform: translate(0px, -5px) scale(9, 2);
            }
            to {
              transform: translate(3.3px, -3.3px) rotate(45deg) scale(5.75, 2);
            }
          }

          @keyframes mabab-b {
            from {
              transform: scale(9, 2);
            }
            to {
              transform: translate(0.25px, 0px) scale(7.5, 2);
            }
          }

          @keyframes mabab-c {
            from {
              transform: translate(0px, 5px) scale(9, 2);
            }
            to {
              transform: translate(3.3px, 3.3px) rotate(-45deg) scale(5.75, 2);
            }
          }

          @keyframes mabab-v {
            from {
              visibility: hidden;
            }
            to {
              visibility: visible;
            }
          }

          .mabab .w,
          .mabab .a,
          .mabab .b,
          .mabab .c {
            animation-duration: var(--delightful-icons-duration);
            animation-fill-mode: forwards;
            animation-timing-function: var(--delightful-icons-easing);
          }

          .mabab .w {
            animation-name: mabab-w;
          }

          .mabab .a {
            animation-name: mabab-a;
          }

          .mabab .b {
            animation-name: mabab-b;
          }

          .mabab .c {
            animation-name: mabab-c;
          }

          .mabab path {
            animation-name: mabab-v;
            animation-delay: var(--delightful-icons-duration);
            animation-fill-mode: forwards;
            visibility: var(--hidden, hidden);
          }
        </style>
        <g transform="translate(12, 12)" stroke="currentcolor">
          <g class="w">
            <line class="a" x1="-1" y1="0" x2="1" y2="0"></line>
            <line class="b" x1="-1" y1="0" x2="1" y2="0"></line>
            <line class="c" x1="-1" y1="0" x2="1" y2="0"></line>
          </g>
        </g>
        <path d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"></path>
      </g>

      <g id="menu" class="mabm">
        <style type="text/css">
          @keyframes mabm-w {
            from {
              transform: rotate(0deg);
              visibility: visible;
            }
            to {
              transform: rotate(180deg);
              visibility: hidden;
            }
          }

          @keyframes mabm-a {
            from {
              transform: translate(-3.3px, -3.3px) rotate(-45deg) scale(5.75, 2);
            }
            to {
              transform: translate(0px, -5px) scale(9, 2);
            }
          }

          @keyframes mabm-b {
            from {
              transform: translate(0.25px, 0px) scale(7.5, 2);
            }
            to {
              transform: scale(9, 2);
            }
          }

          @keyframes mabm-c {
            from {
              transform: translate(-3.3px, 3.3px) rotate(45deg) scale(5.75, 2);
            }
            to {
              transform: translate(0px, 5px) scale(9, 2);
            }
          }

          @keyframes mabm-v {
            from {
              visibility: hidden;
            }
            to {
              visibility: visible;
            }
          }

          .mabm .w,
          .mabm .a,
          .mabm .b,
          .mabm .c {
            animation-duration: var(--delightful-icons-duration);
            animation-fill-mode: forwards;
            animation-timing-function: var(--delightful-icons-easing);
          }

          .mabm .w {
            animation-name: mabm-w;
          }

          .mabm .a {
            animation-name: mabm-a;
          }

          .mabm .b {
            animation-name: mabm-b;
          }

          .mabm .c {
            animation-name: mabm-c;
          }

          .mabm path {
            animation-name: mabm-v;
            animation-delay: var(--delightful-icons-duration);
            animation-fill-mode: forwards;
            visibility: var(--hidden, hidden);
          }
        </style>
        <g transform="translate(12, 12)" stroke="currentcolor">
          <g class="w">
            <line class="a" x1="-1" y1="0" x2="1" y2="0"></line>
            <line class="b" x1="-1" y1="0" x2="1" y2="0"></line>
            <line class="c" x1="-1" y1="0" x2="1" y2="0"></line>
          </g>
        </g>
        <path d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"></path>
      </g>
    </defs>
  </svg>
</iron-iconset-svg>`;

document.head.appendChild($_documentContainer.content);

/**
@license
MIT License

Copyright (c) 2017 Andrew Mitchell

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
/*
  FIXME(polymer-modulizer): the above comments were extracted
  from HTML and may be out of place here. Review them and
  then delete this comment!
*/
;


delightful-icon.js
/**
@license
MIT License

Copyright (c) 2017 Andrew Mitchell

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
/**
`<delightful-icon>` is a helper element that wraps any element that uses an `<iron-icon>`. You can
use it to define which icons to transition to.

Example:

    <delightful-icon icon="menu" next="arrow-back" toggle-on="click">
      <paper-icon-button></paper-icon-button>
    </delightful-icon>

### Styling

The following custom properties are available for styling:

Custom Property               | Description         | Default
------------------------------|---------------------|----------
`--delightful-icons-duration` | Transition duration | `0.2s`
`--delightful-icons-easing`   | Transition easing   | `cubic-bezier(0.4, 0, 0.2, 1)`

@demo demo/index.html
*/
/*
  FIXME(polymer-modulizer): the above comments were extracted
  from HTML and may be out of place here. Review them and
  then delete this comment!
*/
import { PolymerElement, html } from '@polymer/polymer/polymer-element.js';

import { IronMeta } from '@polymer/iron-meta/iron-meta.js';
/**
 * `delightful-icon`
 * Transition between icons using Material Design&#39;s delightful details.
 *
 * @customElement
 * @polymer
 */
class DelightfulIcon extends PolymerElement {
  static get template() {
    return html`
    <style>
      :host([disable-transition]) {
        --delightful-icons-duration: 1ms; /* Safari disables animations at 0s */
      }
    </style>
    <slot></slot>
`;
  }

  static get is() { return 'delightful-icon'; }
  static get properties() {
    return {
      /**
       * The current icon name. Refer to https://material.io/icons/ for icon names. Changing
       * this property will trigger a transition.
       */
      icon: {
        type: String,
        notify: true,
        reflectToAttribute: true,
        observer: 'iconChanged'
      },
      /**
       * The next icon name to transition to when toggling.
       */
      next: {
        type: String,
        notify: true,
        reflectToAttribute: true,
        observer: 'nextChanged'
      },
      /**
       * Optional event name to listen to on the wrapped icon element. When this event fires,
       * the icon will toggle, provided next is set.
       */
      toggleOn: {
        type: String,
        observer: 'toggleOnChanged'
      },
      /**
       * CSS query to locate the wrapped icon element.
       */
      query: {
        type: String,
        value: '*'
      },
      /**
       * The property name to set on the wrapped icon element. This property should consume an
       * <iron-iconset-svg> icon name.
       */
      property: {
        type: String,
        value: 'icon'
      },
      /**
       * Disables icon transitions.
       */
      disableTransition: {
        type: Boolean,
        value: false,
        reflectToAttribute: true
      },
      /**
       * Disables the next icon transition. This will be true when the icon is first set to
       * prevent transitions when the icon loads.
       */
      disableNextTransition: {
        type: Boolean,
        value: true
      },
      /**
       * The wrapped icon element. This element should expose a property that consumes an
       * <iron-iconset-svg> icon name.
       */
      iconElement: {
        type: Object,
        computed: 'getIconElement(query)',
        observer: 'iconElementChanged'
      },
      /**
       * IronMeta instance used to retrieve iconset values.
       */
      _meta: {
        type: Object,
        value: () => {
          return new IronMeta({ type: 'iconset' });
        }
      },
      /**
       * Cached array of the current iconset's fully qualified icon names.
       */
      _icons: {
        type: Array,
        value: () => {
          return [];
        }
      },
      /**
       * Cached array of the current iconset's abbreviated icon names.
       */
      _iconNames: {
        type: Array,
        value: () => {
          return [];
        }
      }
    };
  }

  constructor() {
    super();
    this._toggleOnListener = this.toggle.bind(this);
  }

  /**
   * Toggles between the current icon and the next icon.
   */
  toggle() {
    if (this.next) {
      this.icon = this.next;
    }
  }

  /**
   * Retrieves the wrapped icon element.
   *
   * @return {HTMLElement=} The wrapped icon element.
   */
  getIconElement() {
    return this.querySelector(this.query || '*');
  }

  /**
   * Triggers a transition to the provided icon from a previous icon.
   *
   * @param {string=} icon - The icon to transition to.
   * @param {string=} previous - Optional icon transitioning from.
   */
  iconChanged(icon, previous) {
    let fullIcon = this._icons[this._iconNames.indexOf(icon)];
    if (!fullIcon) {
      if (!this.updatePairFor(icon, previous || this.next)) {
        this.disableNextTransition = true;
        this.updatePairFor(icon);
      }

      if (this._iconNames.indexOf(icon) > -1) {
        this.iconChanged(icon, previous);
      }
    } else if (this.iconElement && this.iconElement[this.property] !== fullIcon) {
      let duration;
      if (this.disableNextTransition) {
        duration = this.iconElement.getComputedStyleValue('--delightful-icons-duration');
        this._previousDisableTransition = this.disableTransition;
        this.disableTransition = true;
      }

      this.iconElement[this.property] = fullIcon;
      this.next = icon === this._iconNames[0] ? this._iconNames[1] : this._iconNames[0];
      if (this.disableNextTransition) {
        setTimeout(() => {
          this.disableTransition = this._previousDisableTransition;
          this.disableNextTransition = false;
        }, this.durationToMs(duration));
      }
    }
  }

  /**
   * Triggered when the next icon changes. When the icon is toggled, the icon is set to next and
   * next is set to the previous icon.
   *
   * @param {string=} next - The current next icon.
   * @param {string=} previous - The previous value of the next icon.
   */
  nextChanged(next, previous) {
    if (next === this.icon) {
      this.icon = previous;
    }
  }

  /**
   * Triggered when toggleOn changes. It will remove the previous event name's listener and add
   * a new listener to the wrapped icon element.
   *
   * @param {string=} toggleOn - The event name to toggle on.
   * @param {string=} previous - The previous event name.
   */
  toggleOnChanged(toggleOn, previous) {
    if (previous && this.iconElement) {
      this.iconElement.removeEventListener(previous, this._toggleOnListener);
    }

    if (toggleOn && this.iconElement) {
      this.iconElement.addEventListener(toggleOn, this._toggleOnListener);
    }
  }

  /**
   * Triggered when iconElement changes. This will remove any toggleOn listeners and ensure
   * listeners are added to the new element.
   *
   * @param {HTMLElement=} element - The current wrapped icon element.
   * @param {HTMLElement=} previous - The previous wrapped icon element.
   */
  iconElementChanged(element, previous) {
    if (previous && this.toggleOn) {
      previous.removeEventListener(this.toggleOn, this._toggleOnListener);
      this.toggleOnChanged(this.toggleOn);
    }
  }

  /**
   * Converts a CSS transition-duration property to milliseconds.
   *
   * @param {string} duration - CSS duration property.
   * @return {number} The number of milliseconds equal to the duration.
   */
  durationToMs(duration) {
    let value = parseFloat(duration);
    if (typeof duration === 'string' && duration.indexOf('ms') === -1) {
      value *= 1000;
    }

    if (isNaN(value)) {
      value = 0;
    }

    return value;
  }

  /**
   * Searches for and updates the iconset transition currently used based on the provided
   * new icon and previous icon.
   *
   * If an iconset cannot be found from the provided icons, a console warning will be generated.
   *
   * @param {string=} newIcon - The icon to transition to.
   * @param {string=} previousIcon - The icon to transition from.
   * @return {boolean} True if an iconset was found, or false if an iconset could not be found.
   */
  updatePairFor(newIcon, previousIcon) {
    let iconset;
    if (newIcon && newIcon.indexOf(':') > -1) {
      iconset = newIcon.split(':')[0];
    }

    if (!iconset) {
      newIcon = this.getIconName(newIcon);
      previousIcon = this.getIconName(previousIcon);
      if (previousIcon) {
        Object.keys(IronMeta.types.iconset || {}).some(key => {
          if (previousIcon && (key === `delightful-${newIcon}-${previousIcon}` ||
              key === `delightful-${previousIcon}-${newIcon}`)) {
            iconset = key;
            return true;
          }
        });
      }

      if (!iconset) {
        Object.keys(IronMeta.types.iconset || {}).some(key => {
          if (key.indexOf('delightful') === 0 && key.indexOf(newIcon) > -1) {
            iconset = key;
            return true;
          }
        });
      }

      if (!iconset) {
        if (previousIcon) {
          console.warn(`Missing icon pair for "${previousIcon}" and "${newIcon}"`);
        } else {
          console.warn(`Missing icon pair for "${newIcon}"`);
        }
      }
    }

    this._meta.key = iconset;
    this._icons = this._meta.value && this._meta.value.getIconNames() || [];
    this._iconNames = this._icons.map(fullIcon => this.getIconName(fullIcon));
    return !!this._meta.value;
  }

  /**
   * Retrieves the abbreviated icon name of a fully qualified icon name.
   *
   * Ex: "av:play-arrow" -> "play-arrow"
   *
   * @param {string=} icon - The abbreviated or fully qualified icon name.
   * @return {string=} The abbreviated icon name.
   */
  getIconName(icon) {
    if (icon && icon.indexOf(':') > -1) {
      return icon.split(':')[1];
    } else {
      return icon;
    }
  }
}

window.customElements.define(DelightfulIcon.is, DelightfulIcon);
